feat: add contract and test

N13e opened the pull request:
Контракт "DomainRegistry" включає такі ключові елементи:
- Реєстрація домену: Функція registerDomain дозволяє контролеру домену реєструвати новий домен за відповідну плату.
- Вивільнення домену: Функція releaseDomain дозволяє контролеру вивільнити зарезервований домен і повернути депозит.
- Поточна інформація про домен: Функція getDomainInfo дозволяє отримати поточну інформацію про домен.
- Поточна інформація про кількість зареєстрованих доменів: Функція getRegisteredDomainsCount дозволяє отримати поточну інформацію по кілкостьсть зареєстрованих доменів.
- Поточна інформація про всі зареєстровані домени: Функція getRegisteredDomains дозволяє отримати поточну інформацію про всі зареєстровані домени. 
- Поточна інформація про всі зареєстровані домени за конкретною адресою: Функція getDomainsForController дозволяє отримати поточну інформацію про всі зареєстровані домени за конкретним адресом контролера. 
- Модифікатор isValidDomain: Цей модифікатор перевіряє, чи є вхідний домен допустимим для реєстрації.
- Події DomainRegistered та DomainReleased: Ці події сповіщають про реєстрацію та вивільнення домену.

N13e committed:
feat: add contract and test

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
+        emit DomainReleased(domain, msg.sender, deposit);
+    }
 
+    function getDomainInfo(string memory domain) external view returns (address controller, uint256 deposit, bool isRegistered) {
```
Why not to return the struct itself?

N13e replied:
Исправлено, таким образом сохраняем всегда целостность структуры. Но деплой контракта с новой реализацией функции дороже. Первая реализация функции - gas 759683, вторая реализация - gas 785841

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
+    struct Domain {
+        address controller;
+        uint256 deposit;
+        bool isRegistered;
```
Do you really need `isRegistered` this field? -1

N13e replied:
Да, можно и без "isRegistered", исправлено.

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
+pragma solidity ^0.8.20;
 
 contract DomainRegistry {
+    uint256 public constant reservationCost = 100; // 0.01 ETH in wei
```
Is it really 0.01 ETH as is in the comment? Can you make this statement more clear in general to understand how much to deposit in ETH? -1

N13e replied:
Исправлено, старый комментарий. Remix не позволяет указывать VALUE в десятичными  числа типа 0.01 и т.д. поэтому игрался со стоимостью

AndriianChestnykh commented on test/DomainRegistry.test.js:
```diff
@@ -1,7 +1,50 @@
+const { expect } = require('chai');
```
Are you sure the tests covered every possible scenarios? -3

N13e replied:
Добавлены еще сценарии

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
+
+    mapping(string => Domain) private domains;
+
+    event DomainRegistered(string domain, address indexed controller, uint256 deposit);
+    event DomainReleased(string domain, address indexed controller, uint256 deposit);
```
You can get info regarding specific controller operations but what about specific domain? This relates to the data monitoring question.

N13e replied:
Информация для мониторинга, очень интересная, ее бы как-то логировать и сохранять.

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
 // SPDX-License-Identifier: UNLICENSED
```
Are you sure the smart contract functionality covers all highly probable questions the business may ask in terms of data monitoring? -3

N13e replied:
В первой реализации были некоторые методы из приведенных ниже, но посчитал, что они не нужны:
- получить весь список зарегистрированных доменов 
- получить весь список доменов, принадлежащих адресу

Потребуется больше места для хранения информации и создание отдельных маппингов для быстрого поиска.

N13e replied:
Добавил методы

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,57 @@
+    }
+
+    function releaseDomain(string memory domain) external {
+        require(domains[domain].controller == msg.sender, "You are not the controller of this domain");
+
+        address controller = domains[domain].controller;
+        uint256 deposit = domains[domain].deposit;
+
+        domains[domain] = Domain({
+            controller: address(0),
+            deposit: 0,
+            isRegistered: false
+        });
+
+        payable(controller).transfer(deposit);
+
+        emit DomainReleased(domain, msg.sender, deposit);
```
Can this code be simplified? -1

N13e replied:
Конечно не оптимален, исправлено.

AndriianChestnykh reviewed (commented):
-9 points in total.
Feel free to fix

N13e committed:
feat: fix getDomainInfo

N13e committed:
feat: fix isValidDomain and cleanups

N13e committed:
feat: fix reservationCost comment

N13e committed:
feat: add some tests

N13e committed:
feat:
add methods: getDomainsForController
 getRegisteredDomainsCount
 getRegisteredDomains

add removeDomainFromList

N13e committed:
feat: fix and cleanups

N13e committed:
feat: add test for getRegisteredDomainsCount

N13e committed:
feat: add test for getRegisteredDomains

N13e committed:
feat: fix test for getRegisteredDomains

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,85 @@
+pragma solidity ^0.8.20;
 
 contract DomainRegistry {
+    uint256 public constant reservationCost = 100; // 0.0000000001 Ether or 0.0000001 ETH in wei
```
Your comment is wrong. Do you really want to set (10^-16) fraction of Ether as a collateral for domain? Does it really work as spam protection?
-1 point

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,85 @@
+    }
+
+    mapping(string => Domain) private domains;
+    string[] private domainNames;  // Array to store domain names
```
Your comment is for obvious code line here and in many other places
-1 point

AndriianChestnykh commented on test/DomainRegistry.test.js:
```diff
@@ -1,7 +1,115 @@
+
+    // Check domain info
+    const domainInfo = await domainRegistry.getDomainInfo(domain);
+    expect(domainInfo.controller).to.equal('0x0000000000000000000000000000000000000000');
```
you have ZeroAddress in `ethers.js`

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,85 @@
+    }
+
+    function getDomainsForController(address controller) external view returns (string[] memory) {
+        string[] memory controllerDomains = new string[](domainNames.length);
+        uint256 count = 0;
+        for (uint256 i = 0; i < domainNames.length; i++) {
+            if (domains[domainNames[i]].controller == controller) {
+                controllerDomains[count] = domainNames[i];
+                count++;
+            }
+        }
+        return controllerDomains;
```
1) What if you would have hundreds of thousands domains in your contract. What would be performance of such a function?
-1 point
2) Why `string[] memory` length can be greater than actual data you return? It's length is equal to all domains quantity but not to the quantity which belong to special controller.
-1 point

AndriianChestnykh commented on contracts/DomainRegistry.sol:
```diff
@@ -1,9 +1,85 @@
+    }
 
+    function removeDomainFromList(string memory domain) internal {
+        for (uint256 i = 0; i < domainNames.length; i++) {
+            if (keccak256(abi.encodePacked(domainNames[i])) == keccak256(abi.encodePacked(domain))) {
+                if (i != domainNames.length - 1) {
+                    domainNames[i] = domainNames[domainNames.length - 1];
+                }
+                domainNames.pop();
+                break;
+            }
+        }
+    }
```
The same. What would be performance and, event more important, gas cost of such a function if you have hundreds of thousands domains? Not saying that you may just reach 30M gas limit for a block in some blockchain network.
-2 points

AndriianChestnykh reviewed (commented):
This implementation is much better in terms of data monitoring. Although its performance and gas consumption is far from optimal.
-6 points in total

N13e committed:
feat: fix tests and cleanups

N13e committed:
feat: fix prices in contract and tests

N13e committed:
feat: fix tests

N13e committed:
feat: fix tests
